sroka/api/mysql/mysql.py

```diff
@@ -0,0 +1,124 @@
+from configparser import NoOptionError, NoSectionError
+
+import os
+import mysql.connector
+from mysql.connector.errors import DatabaseError, OperationalError, InternalError
+import pandas as pd
+from retrying import retry
+
+import sroka.config.config as config
+
+
+@retry(stop_max_attempt_number=1,
+       wait_exponential_multiplier=1 * 2,
+       wait_exponential_max=1 * 2 * 2)
+def query_mysql(query: str, filename=None):
+    try:
+        options = get_options_from_config()
+
+    except NoSectionError:
+        print('Missing MySQL section in configuration')
+        return pd.DataFrame([])
+
+    if not validate_options(options):
+        return pd.DataFrame([])
+
+    try:
+        # Connect while passing only the arguments that were set in the
+        # configuration, since some are mutually exclusive or optional.
+        connection = mysql.connector.connect(**{k: v for k, v in options.items() if v != ""})
+
+        # Get the MySQL connector cursor, which allows interacting with the
+        # MySQL server.
+        cursor = connection.cursor()
+
+        # Execute the query.
+        cursor.execute(query)
+
+    except OperationalError as e:
+        print('Operational MySQL Error: {}'.format(e))
+        return pd.DataFrame([])
+    except InternalError as e:
+        print('Internal MySQL Error: {}'.format(e))
+        return pd.DataFrame([])
+    except DatabaseError as e:
+        print('Database MySQL Error: {}'.format(e))
+        return pd.DataFrame([])
+
+    # Cycle through the returned elements to build a pandas DataFrame.
+    df = pd.DataFrame(cursor, columns=cursor.column_names)
+
+    # Close the connection to the MySQL server.
+    cursor.close()
+    connection.close()
+
+    # Get path to folder in which to create the file.
+    filepath = "\\".join(filename.split('\\')[:-1])
```

Comment:
filename can be None (if not defined). In such case it throws an error `'NoneType' object has no attribute 'split'` we need to do it only if we want to save data to file.

Reply:
btw I think it would be worth considering for both windows and unix notations.

+ if user defines a path without escaping backslashes so e.g. "test\test_file.csv" it will not work. We could in such case use raw string. but what if user defines string with "\\"...

how about using `pathlib` for it?

Reply:
Ah, I thought this solution would work for every OS, my bad! I will look into this more then :)

Reply:
Yeah, I would say the main issue is assuming that people will define path with \\\\

Currently, this works:
```
import pandas as pd
from sroka.api.mysql.mysql import query_mysql

df = query_mysql("SELECT * FROM clan LIMIT 10", "test\\test_file.csv")
df = pd.read_csv("test\\test_file.csv")
df.head()
```

but those doesn't:
```
import pandas as pd
from sroka.api.mysql.mysql import query_mysql

df = query_mysql("SELECT * FROM clan LIMIT 10", "test/test_file.csv")
df = pd.read_csv("test/test_file.csv")
df.head()
```

```
import pandas as pd
from sroka.api.mysql.mysql import query_mysql

df = query_mysql("SELECT * FROM clan LIMIT 10", "test\test_file.csv")
df = pd.read_csv("test\test_file.csv")
df.head()
```

Because of split definition.

Let us know if you need any support with this.

Reply:
As you suggested, I just pushed a version that uses Python's native `pathlib` and it seems to work fantastically well :) Let me know what you think, and thanks for the great reviews! 👍